lib/OpenQA/WebAPI/Plugin/AMQP.pm

```diff
@@ -69,6 +70,178 @@ sub log_event {
+    # anyone could adopt it
+    my ($self, $event, $job, $baseurl) = @_;
+    my $stdevent;
+    my $clone_of;
+    my $job_id;
+    # first, get the standard 'state' (from 'queued', 'running',
+    # 'complete', 'error'; we cannot do 'running' at present
+    if ($event eq 'openqa_job_create') {
+        $stdevent = 'queued';
+        $job_id   = $job->id;
+    }
+    elsif ($event eq 'openqa_job_restart' || $event eq 'openqa_job_duplicate') {
+        $stdevent = 'queued';
+        $clone_of = $job->id;
+        $job_id   = $job->clone_id;
+    }
+    elsif ($event eq 'openqa_job_cancel') {
+        $stdevent = 'error';
+        $job_id   = $job->id;
+    }
+    elsif ($event eq 'openqa_job_done') {
+        $job_id = $job->id;
+        # lifecycle note: any job cancelled directly via the web API will
+        # see both job_cancel and job_done with result USER_CANCELLED, so
+        # we emit duplicate standardized fedmsgs in this case. This is
+        # kinda unavoidable, though, as it's possible for a job to wind up
+        # USER_CANCELLED *without* an openqa_job_cancel event happening,
+        # so we can't just throw away all openqa_job_done USER_CANCELLED
+        # events...
+        if (grep { $job->result eq $_ } OpenQA::Schema::Result::Jobs::INCOMPLETE_RESULTS) {
+            $stdevent = 'error';
+        }
+        else {
+            $stdevent = 'complete';
+        }
+    }
+    else {
+        return;
+    }
+
+    # we need this for the system dict; it should be the release of
+    # the system-under-test (the VM in which the test runs) at the
+    # *start* of the test, I think. We're trying to capture info about
+    # the environment in which the test runs
+    my $sysrelease = $job->VERSION;
+    my $hdd1;
+    my $bootfrom;
+    $hdd1       = $job->settings_hash->{HDD_1}    if ($job->settings_hash->{HDD_1});
+    $bootfrom   = $job->settings_hash->{BOOTFROM} if ($job->settings_hash->{BOOTFROM});
+    $sysrelease = $1                              if ($hdd1 && $hdd1 =~ /disk_f(\d+)/ && $bootfrom eq 'c');
+
+    # next, get the 'artifact' (type of thing we tested)
+    my $artifact;
+    my $artifact_id;
+    my $artifact_release;
+    my $compose_type;
+    my $test_namespace;
+    # current date/time in ISO 8601 format
+    my $generated_at = strftime("%Y-%m-%dT%H:%M:%S", gmtime()) . 'Z';
+
+    my $build = $job->BUILD;
+    if ($build =~ /^Fedora/) {
+        $artifact       = 'productmd-compose';
+        $artifact_id    = $build;
+        $compose_type   = 'production';
+        $compose_type   = 'nightly' if ($build =~ /\.n\./);
+        $compose_type   = 'test' if ($build =~ /\.t\./);
+        $test_namespace = 'compose';
+    }
+    elsif ($build =~ /^Update-/) {
+        $artifact    = 'fedora-update';
+        $artifact_id = $build;
+        $artifact_id =~ s/^Update-//;
+        $artifact_release = $job->VERSION;
+        $test_namespace   = 'update';
+    }
```

Comment:
ok, this part is actually *very* product-specific for openQA "upstream" code.

Reply:
yeah, it's already in the fedmsg plugin, though. :D that is kinda why I made this configurable, too.

Reply:
I tweaked this a bit to match on `Update-FEDORA` instead, which makes it clearer that it's very Fedora-ish and also makes it a bit more obvious that at least theoretically you could have different conditionals for non-Fedora 'artifacts' here, if this message spec were to take off wildly and become an industry-wide sensation or something.